### "You built a spam filter. It looked at 100 emails. Write down what FOUR things could happen for each email."

> - I have built the model, now the model filter spam mails which means it will go through the normal mails and it will detect and filters out the mail which spam. So when the model is filter out there will be certain things which will evantually done by the model. One of those will be like it will predict the spam mail correctly which means (the mail belongs to spam mail and model filter predicts it is spam mail), second thing will be like, you got mail which is not a spam mail and model filter predicts it is spam filter which is not true and dangerous because what if the mail is important one to user and model predicts as spam filter. and other things which can happen is the mail is spam only but model predicts it is not spam, in which model is not doing it's job and evantually user inbox fills with spam mails. and last thing which happens is, user got a mail which is not spam and model also says that it is not spam.
               actual     |       model
mail(spam)       T        |         T
                 F        |         T
                 T        |         F
                 F        |         F

                 
           

### "If I told you accuracy is 95%, why might that be misleading? Invent a scenario with numbers."

> - What i will get when someone says 95% accuracy, so consider you are doing the experiment for 100times, in this model suppose you we are doing spam filter, when you send 100 mails one by one 95times it will correctly whether the mail is spam or not.

> - if a given data set 95 out of 100 are belong to class 0, so when model predicts all are 0. it is misleading in a way that data itself having so much noise and every time it predicts 0 because of nature of the dataset and it is misleading.
                


Call answers the question: How many of them correctly identified? (actual spam mails are 20, then call gives 10, TPR will be 50% because out of 20 spam mails only 10 of them were identified as spam)
Precision answers the question: How many are correctly classified as positives? (model predicted 10 spam mails out of 20 spam mails and all of identified spam mails are spam only then precision becomes 100%)

100 mails are there...

actual:
20 - spam mails 
80 - not spam mails

model predicted:

10 - spam mails (TP)
10 - not spam (FN)
80 - not spam mails (TN)

TPR:

10 / (10 + 10) = 0.5% (50 %)


Precision :
precision = 10 / (10 + 0) = 100%

In [3]:
y_true = [1, 0, 1, 1, 0, 0, 1, 0, 1, 1]
y_pred = [1, 0, 0, 1, 0, 1, 1, 0, 0, 1]

                Predicted spam      Predicted not-spam
Actual spam           4                     2
Actual not-spam       1                     3

TP : 4
FP : 1
TN : 3
FN : 2

TPR = TP / (TP + FN) = 4/6 = 0.67
Precision = TP / (TP+FP) = 4/4+1 = 4/5 = 0.8


In [10]:
from sklearn.metrics import confusion_matrix, classification_report, roc_auc_score
from sklearn.datasets import make_classification
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
import numpy as np

In [5]:
tn, fp, fn, tp = confusion_matrix(y_true=y_true, y_pred=y_pred).ravel().tolist()
print(f'tp: {tp}, fp: {fp}, tn: {tn}, fn: {fn} ')

print(classification_report(y_true,y_pred))

tp: 4, fp: 1, tn: 3, fn: 2 
              precision    recall  f1-score   support

           0       0.60      0.75      0.67         4
           1       0.80      0.67      0.73         6

    accuracy                           0.70        10
   macro avg       0.70      0.71      0.70        10
weighted avg       0.72      0.70      0.70        10



In [6]:
X, y = make_classification(n_samples=200, n_features=2, random_state=42, n_redundant=0, n_clusters_per_class=1)

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, test_size=0.2)

model = LogisticRegression()
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

tn, fp, fn, tp = confusion_matrix(y_test, y_pred).ravel().tolist()
print(f' TP: {tp}\n FP: {fp}\n TN: {tn}\n FN: {fn} ')

 TP: 17
 FP: 5
 TN: 18
 FN: 0 


In [7]:
y_proba  = model.predict_proba(X_test)
print(y_proba.shape)

print(y_proba[:5])

(40, 2)
[[0.90193467 0.09806533]
 [0.30218044 0.69781956]
 [0.29856701 0.70143299]
 [0.36220685 0.63779315]
 [0.18991482 0.81008518]]


Felt overwhelmed with the concept and learning for the day, will comeback tomorrow freshly and finish what i left in tomorrow morning. will spent atleast 30min. to finish the mini task that started.

Total hours logged on Monday: 2hrs approx.

In [15]:
positive_class_proba = y_proba[:,1]
print(positive_class_proba)


auc = roc_auc_score(y_test, positive_class_proba)
print(f"ROU-AUC: {auc:.3f}")

[0.09806533 0.69781956 0.70143299 0.63779315 0.81008518 0.87985207
 0.06816319 0.24620286 0.37486239 0.06116108 0.77010988 0.00494304
 0.99237741 0.88145073 0.45075596 0.72488504 0.03274399 0.13950633
 0.95803992 0.81387601 0.9511782  0.77958244 0.88641623 0.15063383
 0.05596599 0.37650275 0.86584472 0.18734075 0.64803683 0.9587097
 0.35340091 0.83896055 0.17797413 0.21973875 0.58911869 0.30884523
 0.93181967 0.01796455 0.59668511 0.50651377]
ROU-AUC: 0.934


## If I had passed y_pred (0s and 1s) instead of positive_class_proba into roc_auc_score, what would I lose and why?

> ROC-AUC is nothing but model evaluates the ranking quality based on ranking positive lables more than negative label. so when you give 0's and 1's instead of probabilities it will destroy the model ranking because passing 0's and 1's will reduce threshold descision.

In [36]:
threshold = 0.3

y_pred_custom = (positive_class_proba >= threshold).astype(int)

print(f"Threshold: {threshold}")
tn, fp, fn, tp =  confusion_matrix(y_test, y_pred_custom).ravel().tolist()

print(f'recall: {tp/ (tp + fn):.2f}')
print(f'precision: {tp/(tp+fp):.2f}')
print(confusion_matrix(y_test, y_pred_custom))
print(roc_auc_score(y_test, y_pred_custom))


Threshold: 0.3
recall: 1.00
precision: 0.63
[[13 10]
 [ 0 17]]
0.782608695652174


For threshold 0.3, recall is 1.00 which means that it predicted all tp(number of tps) and precision says only 63% are correct. 

In [34]:
threshold = 0.5

y_pred_custom = (positive_class_proba >= threshold).astype(int)

print(f"Threshold: {threshold}")
tn, fp, fn, tp =  confusion_matrix(y_test, y_pred_custom).ravel().tolist()

print(f'recall: {tp/ (tp + fn):.2f}')
print(f'precision: {tp/(tp+fp):.2f}')
print(confusion_matrix(y_test, y_pred_custom))
print(roc_auc_score(y_test, y_pred_custom))

Threshold: 0.5
recall: 1.00
precision: 0.77
[[18  5]
 [ 0 17]]
0.891304347826087


For threshold 0.5, recall is 1.00 which means that it predicted all tp(number of tps) and precision says only 77% are correct. 

In [35]:
threshold = 0.7

y_pred_custom = (positive_class_proba >= threshold).astype(int)

print(f"Threshold: {threshold}")
tn, fp, fn, tp =  confusion_matrix(y_test, y_pred_custom).ravel().tolist()

print(f'recall: {tp/ (tp + fn):.2f}')
print(f'precision: {tp/(tp+fp):.2f}')
print(confusion_matrix(y_test, y_pred_custom))
print(roc_auc_score(y_test, y_pred_custom))

Threshold: 0.7
recall: 0.76
precision: 0.81
[[20  3]
 [ 4 13]]
0.8171355498721228


For threshold 0.3, recall is 0.7 which means that it predicted 70% of tp and precision says only 81% are correct. 

## Who controls the threshold — the model or you?
> - The threshold will be controlled by me because based on the recall, precision and ROC_AUC score it will be decided by me based on the tradeoff which is like if you want more precision then the you will be trading off the number of tp's based on the recall.